fusesoc/build/vivado.py

```diff
@@ -0,0 +1,190 @@
+
+        ipconfig = '\n'.join(['read_ip '+s for s in ip])+"\n"
+        ipconfig += "upgrade_ip [get_ips]\n"
+        ipconfig += "generate_target all [get_ips]\n"
+
+        parameters = ""
+        for key, value in self.vlogparam.items():
+            parameters += "set_property generic {{{key}={value}}} [current_fileset -simset]".format(key=key, value=value)
+
+        # Write the formatted string to the tcl file
+        tcl_file.write(PROJECT_TCL_TEMPLATE.format(
+            design       = self.system.sanitized_name,
+            part         = self.system.system.backend.part,
+            bitstream    = os.path.join(self.work_root, self.system.sanitized_name+'.bit'),
+            incdirs      = ' '.join(incdirs),
+            ip           = ipconfig,
+            parameters   = parameters,
+            src_files    = '\n'.join(['read_verilog '+s for s in verilog]+
+                                     ['read_vhdl '+s for s in vhdl]),
+            xdc_files    = '\n'.join(['read_xdc '+s for s in constr])))
+
+        tcl_file.close()
+
+    """ Execute the build
+
+    This launches the actual build of the vivado project by executing the project
+    tcl file in batch mode.
+    """
+    def build(self, args):
+        super(Vivado, self).build(args)
+
+        utils.Launcher('vivado', ['-mode', 'batch', '-source',
+                                  os.path.join(self.work_root, self.system.sanitized_name+'.tcl')],
+                       cwd = self.work_root,
+                       errormsg = "Failed to build FPGA bitstream").run()
+
+        super(Vivado, self).done()
+
+    """ Program the FPGA
+
+    For programming the FPGA a vivado tcl script is written that searches for the
+    correct FPGA board and then downloads the bitstream. The tcl script is then
+    executed in Vivado's batch mode.
+    """
+    def pgm(self, remaining):
+        tcl_file_name = os.path.join(self.work_root, self.system.name+"_pgm.tcl")
+        self._write_program_tcl_file(tcl_file_name)
+        utils.Launcher('vivado', ['-mode', 'batch', '-source', tcl_file_name ],
+                       cwd = self.work_root,
+                       errormsg = "Failed to program the FPGA").run()
+
+    """ Write the programming TCL file """
+    def _write_program_tcl_file(self, program_tcl_filename):
+        tcl_file = open(program_tcl_filename, 'w')
+        tcl_file.write(PROGRAM_TCL_TEMPLATE.format(
+            bitstream    = os.path.join(self.work_root, self.system.name+'.bit'),
+            hw_device = self.system.backend.hw_device
+        ))
+        tcl_file.close()
+
+
+""" Template for vivado project tcl file """
+PROJECT_TCL_TEMPLATE = """# Auto-generated project tcl file
+
+create_project -part {part} {design}
+
+set_property "simulator_language" "Mixed" [current_project]
+
+{ip}
+
+{src_files}
+
+{parameters}
+
+set_property include_dirs [list {incdirs}] [current_fileset]
+
+{xdc_files}
+
+regexp -- {{Vivado v([0-9]{{4}})\.[0-9]}} [version] -> year
```

Comment:
Should add a comment there. The "stragies" are named with the tool version and as we don't want a fixed version in it, we always use the newest. Hence, some tcl magic..